tuist/tuist
# Support xcodebuild arguments in build and test commands

Resolves https://github.com/tuist/tuist/issues/6292

### Short description 📝

As described in the issue, supporting standard xcodebuild arguments in `tuist build` and `tuist test` doesn't scale. For every _first-class_ xcodebuild parameter we want to provide good ergonomics for, we also add some overhead if some arguments are shortened or reworded for clarity, while everybody is familiar with `xcodebuild` arguments already.

Also, consider the amount of boilerplate and validation we do in `XcodeBuildArgument` for parsing the argument into a type and the other way around when passing the arguments to `xcodebuild`. The typed xcodebuild arguments are not justified as they're not used in different command pipeline parts.

### How to test the changes locally 🧐

For `tuist build`, the arguments `clean`, `device`, `platform`, `os`, `rosetta`, `configuration`, and `derivedDataPath` are now deprecated and you will get a warning asking you to use the default `xcodebuild` arguments after the terminator `--`. The same will happen when using `tuist test` with a couple of extra deprecated arguments: `retryCount` and `resultBundlePath`.

```bash
tuist build --path PATH --configuration Debug # warning
```

Running `tuist build` or `tuist test` with arguments already supported by `xcodebuild` (after the terminator `--`) should work as expected:

```bash
tuist build --path PATH -- -configuration Debug
```

Please note that tuist already sets some `xcodebuild` parameters for you. If you try to use the arguments `-scheme`, `-workspace`, or `-project` after the terminator `--` it will throw an error. For `tuist test` the disallowed arguments also includes `-testPlan`, `-skip-test-configuration`, `-only-test-configuration`, `-only-testing`, and `-skip-testing`.

```bash
tuist build -- -project MyApp # error
```

### Contributor checklist ✅


### Reviewer checklist ✅
